crossdock/behavior/trace/behavior.go

```diff
@@ -0,0 +1,236 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package trace
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/uber/tchannel-go"
+
+	"github.com/crossdock/crossdock-go"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/utils"
+	"golang.org/x/net/context"
+)
+
+// Different parameter keys and values used by the system
+const (
+	BehaviorName = "trace"
+)
+
+// Behavior is an implementation of "trace behavior", that verifies
+// that a tracing context and baggage are properly propagated through
+// two level of servers.
+type Behavior struct {
+	ServerPort    string
+	Tracer        opentracing.Tracer
+	ServiceToHost func(string) string
+	ch            *tchannel.Channel
+	thriftCall    DownstreamCall
+	jsonCall      DownstreamCall
+}
+
+// DownstreamCall is used in a few other structs here
+type DownstreamCall func(ctx context.Context, target *Downstream) (*Response, error)
+
+// Register function adds JSON and Thrift handlers to the server channel ch
+func (b *Behavior) Register(ch *tchannel.Channel) {
+	b.registerThriftServer(ch)
+	b.registerJSONServer(ch)
+}
+
+// Run executes the trace behavior
+func (b *Behavior) Run(t crossdock.T) {
+	sampled, err := strconv.ParseBool(t.Param(sampledParam))
+	if err != nil {
+		t.Fatalf("Malformed param %s: %s", sampledParam, err)
+		return
+	}
+	baggage := randomBaggage()
+
+	level1 := &Request{
+		ServerRole: RoleS1,
+	}
+	server1 := t.Param(server1NameParam)
+
+	level2 := &Downstream{
+		ServiceName: t.Param(server2NameParam),
+		ServerRole:  RoleS2,
+		HostPort: fmt.Sprintf("%s:%s",
+			b.serviceToHost(t.Param(server2NameParam)),
+			b.ServerPort,
+		),
+		Encoding: t.Param(server2EncodingParam),
+	}
+	level1.Downstream = level2
+
+	level3 := &Downstream{
+		ServiceName: t.Param(server3NameParam),
+		ServerRole:  RoleS3,
+		HostPort: fmt.Sprintf("%s:%s",
+			b.serviceToHost(t.Param(server3NameParam)),
+			b.ServerPort,
+		),
+		Encoding: t.Param(server3EncodingParam),
+	}
+	level2.Downstream = level3
+
+	resp, err := b.startTrace(level1, sampled, baggage)
+	if err != nil {
+		t.Errorf(err.Error())
+		return
+	}
+
+	traceID := resp.Span.TraceID
+	if traceID == "" {
+		t.Errorf("Trace ID is empty in S1(%s)", server1)
```

Comment:
FYI, if you use the Assertions API, you'll get a stack trace in the error message in case of failure.

https://godoc.org/github.com/crossdock/crossdock-go#Assertions